### PR TITLE
GODRIVER-1235 Skip embedded documents and arrays correctly in extJSONValueReader

### PR DESCRIPTION
**Pull request by benjirewis**
[GODRIVER-1235](https://jira.mongodb.org/browse/GODRIVER-1235) 

Adds logic to correctly skip embedded documents and embedded arrays in `extJSONValueReader` by creating one function to skip both iteratively using the parser's `advanceState()` function. This addresses errors in unmarshalling extended JSON with undefined fields with nested documents or arrays.

Adds tests for unmarshalling extended JSON with undefined fields in `unmarshal_test.go`.

### CODE REVIEW

File: bson/bsonrw/extjson_reader.go [hunk 1]
```diff
@@ -225,7 +230,7 @@ func (ejvr *extJSONValueReader) Type() bsontype.Type {
 
 func (ejvr *extJSONValueReader) Skip() error {
 	switch ejvr.stack[ejvr.frame].mode {
-	case mElement, mValue:
+	case mElement, mValue, mArray:
```
Comment (benjirewis) [inline bson/bsonrw/extjson_reader.go:233]: Per offline discussion, I added a test for skipping an array alongside the embedded document test. `extjson_reader#skipArray()` was actually working just fine, but for some reason this case in `Skip` (which calls `skipArray()`) was missing `mArray` and was throwing the error below when an array was skipped. Should be working now.

**Review by iwysiu — APPROVED** [review]

File: bson/unmarshal_test.go [hunk 2]
```diff
@@ -166,3 +167,86 @@ func TestCachingDecodersNotSharedAcrossRegistries(t *testing.T) {
...
+				"UndefinedField": {"key": {"one": {"two": [3]}}},
+				"DefinedField": "value"
+			}`,
+		},
+		{
+			"embedded document and embedded array in outer array",
+			`{
+				"UndefinedField": [{"key": {"one": [3]}}],
+				"DefinedField": "value"
+			}`,
```
Comment (kevinAlbs) [inline bson/unmarshal_test.go:242]: Nice tests! Can we add cases for BSON values of type `CodeWithScope`, e.g. `{"$code": "foo", "$scope": {"bar": 1}}`? I only ask because `extJSONParser.peekType` and `extJSONParser.readValue` do handle it specially and it has an embedded document.

Also, in case you have not seen it, https://github.com/mongodb/specifications/blob/master/source/extended-json.rst#conversion-table has a nice reference for the forms of each BSON type in extended canonical JSON.

**Review by kevinAlbs — COMMENTED** [review]
Looking good! I just have a question about whether we can use `peekType` to skip documents and request some additional tests.

File: bson/bsonrw/extjson_reader.go [hunk 3]
```diff
@@ -163,11 +163,24 @@ func (ejvr *extJSONValueReader) skipDocument() error {
 	// read entire document until ErrEOD (using readKey and readValue)
 	_, typ, err := ejvr.p.readKey()
 	for err == nil {
-		_, err = ejvr.p.readValue(typ)
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply bson/bsonrw/extjson_reader.go:166]: Yep! The updated `skipObject()` uses only `peekType()`, and the function is now iterative, but the concept is the same.

File: bson/bsonrw/extjson_reader.go [hunk 4]
```diff
@@ -176,9 +189,16 @@ func (ejvr *extJSONValueReader) skipDocument() error {
 
 func (ejvr *extJSONValueReader) skipArray() error {
 	// read entire array until ErrEOA (using peekType)
-	_, err := ejvr.p.peekType()
+	typ, err := ejvr.p.peekType()
 	for err == nil {
-		_, err = ejvr.p.peekType()
+		typ, err = ejvr.p.peekType()
+		// account for embedded arrays
+		if typ == bsontype.Array {
```
[the remark this answers is not in the document]
Comment (benjirewis) [reply bson/bsonrw/extjson_reader.go:196]: Great point! The `skipObject()` function now skips both documents and arrays and uses only `peekType()`.

File: bson/unmarshal_test.go (hunk 2 above)
Comment (benjirewis) [reply bson/unmarshal_test.go:242]: Great catch! Added some `CodeWithScope` tests and a test for embedded documents with type definitions, as in `{"myDouble": {"$numberDouble": "1.24"}}`.

File: bson/bsonrw/extjson_reader.go [hunk 5]
```diff
@@ -159,26 +159,18 @@ func (ejvr *extJSONValueReader) pop() {
...
-	// read entire array until ErrEOA (using peekType)
+func (ejvr *extJSONValueReader) skipObject() error {
+	// read entire object until encountering EOA or EOD, and depth is one level below original
 	_, err := ejvr.p.peekType()
+	initialDepth := ejvr.p.depth
 	for err == nil {
 		_, err = ejvr.p.peekType()
+		if err == ErrEOA || err == ErrEOD {
+			err = nil
+			if ejvr.p.depth == initialDepth-1 {
```
Comment (kevinAlbs) [inline bson/bsonrw/extjson_reader.go:170]: Gah, I pointed you in the wrong direction. Looking at `advanceState`, `depth` only applies to documents, not arrays. Sorry about that!

File: bson/bsonrw/extjson_parser.go [hunk 6]
```diff
@@ -104,6 +104,13 @@ func (ejp *extJSONParser) peekType() (bsontype.Type, error) {
 	case jpsSawEndArray:
 		// this would only be a valid state if we were in array mode, so return end-of-array error
 		err = ErrEOA
+	case jpsSawEndObject:
+		if ejp.peekMode() == jpmObjectMode {
```
Comment (kevinAlbs) [inline bson/bsonrw/extjson_parser.go:108]: Looking over this again, I still feel like this is a little fragile. I think I'd prefer not modifying `peekType` if possible.

The comment on L105 justifies why the `peekMode()` is not checked for the array case. So, I thought it would be safe to remove this condition, but that led me astray.

Here is my rough understanding.

`extJSONParser` maintains a stack of modes for parsing nested arrays and objects. `advanceState` calls `pushMode` when encountering an `[` or `{` token, and calls `popMode` when encountering a `]` or `}` token to check that brackets are balanced and correct.

So when `advanceState` returns `jpsSawEndObject`, the mode stack for the object that was being read was just popped. Consequently, I think this condition is only true when popping a nested object. For example:

```
[ { "a" : { "b": 1 } } ]
```

When reading the first `}` this condition will be true.
But when reading the second `}`, we'll pop the `jpmObjectMode`. The mode stack will contain one `jpmArrayMode`. So this condition is false.

In summary, I don't quite understand the need for the check, and I am not sure why removing this check breaks tests.

There are other oddities about `peekType` (e.g. not all states are handled in this `switch`, so I think `peekType` just returns the 0 type in those cases...).

So unless we have a solid understanding, I'd like to vote for another approach. I think we can bypass `peekType` entirely by using `advanceState` (explained below).

File: bson/bsonrw/extjson_reader.go [hunk 7]
```diff
@@ -159,26 +159,18 @@ func (ejvr *extJSONValueReader) pop() {
...
-
-		_, typ, err = ejvr.p.readKey()
-	}
-
-	return err
-}
-
-func (ejvr *extJSONValueReader) skipArray() error {
-	// read entire array until ErrEOA (using peekType)
+func (ejvr *extJSONValueReader) skipObject() error {
```
Comment (kevinAlbs) [inline bson/bsonrw/extjson_reader.go:162]: I think we can tweak this to only use `advanceState`.

The initial state, `ejvr.p.s` should to be `jpsSawBeginArray` or `jpsSawBeginObject` at the beginning of this function. I think we can do something like:

- initialize `count` to 1
- call advanceState
- if the state is `jpsSawBeginArray` or `jpsSawBeginObject`, increment `count`
- if the state is `jpsSawEndArray` or `jpsSawEndObject`, decrement `count`
- repeat until `count` is 0

**Review by kevinAlbs — COMMENTED** [review]
The iterative approach is nicer! I think we can avoid `peekType` entirely. Let me know if you'd like to discuss.

File: bson/bsonrw/extjson_parser.go (hunk 6 above)
Comment (benjirewis) [reply bson/bsonrw/extjson_parser.go:108]: Great points, and thank you for looking so in-depth into `peekType()`. I think I agree, and I've switched to using `advanceState()` with the algorithm you described below. The only thing we lose is the error-checking in `peekType()`; `advanceState()` advances forward blindly and does not return errors, so now neither does `skipObject()`. There are some disadvantages to this I may bring up offline...

File: bson/bsonrw/extjson_reader.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (benjirewis) [reply bson/bsonrw/extjson_reader.go:166]: Now it uses `advanceState()` instead!

File: bson/bsonrw/extjson_reader.go (hunk 7 above)
Comment (benjirewis) [reply bson/bsonrw/extjson_reader.go:162]: Sounds good! That algorithm seems to work perfectly for all existing examples.

File: bson/bsonrw/extjson_reader.go (hunk 5 above)
Comment (benjirewis) [reply bson/bsonrw/extjson_reader.go:170]: Ah, you're right, all good!

File: bson/bsonrw/extjson_parser.go (hunk 6 above)
Comment (kevinAlbs) [reply bson/bsonrw/extjson_parser.go]: Per offline discussion, I am not too concerned with losing some skip validation on malformed BSON. The previous behavior was incorrect on valid BSON.

**Review by kevinAlbs — APPROVED** [review]
LGTM!

File: bson/unmarshal_test.go [hunk 8]
```diff
@@ -166,3 +167,114 @@ func TestCachingDecodersNotSharedAcrossRegistries(t *testing.T) {
 		assert.Equal(t, int32(-1), *second.X, "expected X value to be -1, got %v", *second.X)
 	})
 }
+
+type expectedResponse struct {
+	DefinedField string
+}
+
+func unmarshalExtJSONHelper(extJSON string) (*expectedResponse, error) {
```
Comment (divjotarora) [inline bson/unmarshal_test.go:175]: nit: I understand the purpose of this, but functions with generic names like this can cause issues for future authors. Someone trying to define a similar function for a different type in a test would run into trouble because the name is already taken here, so we'd end up with more convoluted names.

I'd recommend defining both the `expectedResponse` type and this function within the test function so their scope is limited. I have a small preference for this approach because it makes it very obvious that this type is tied to a specific test:

```
func TestUnmarshalExtJSONWithUndefinedField(t *testing.T) {
    type expectedResponse struct {
        DefinedField string
    }

    unmarshalExpectedResponse := func(t *testing.T, extJSON string) {
        t.Helper()
        err := UnmarshalExtJSON([]byte(extJSON), false, &responseDoc)
        assert.Nil(t, err, "UnmarshalExtJSON error: %v", err)
    }
}
```
Note that I changed the function slightly to give it a different name and have it directly accept a `testing.T` instance so it can fail the test if unmarshalling errors. The `t.Helper()` call at the beginning makes it so stack traces for a failed test will leave this function out (see https://gist.github.com/divjotarora/7ea6f953fd3cd7d21f0bd170e319768b for an example).

File: bson/unmarshal_test.go [hunk 9]
```diff
@@ -166,3 +167,114 @@ func TestCachingDecodersNotSharedAcrossRegistries(t *testing.T) {
...
+			`{
+				"UndefinedField": {"empty": {}},
+				"DefinedField": "value"
+			}`,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			responseDoc, err := unmarshalExtJSONHelper(tc.testJSON)
+			require.NoError(t, err)
```
Comment (divjotarora) [inline bson/unmarshal_test.go:276]: Bit of legacy context: many of our older tests use `testify/require` to do assertions, but we're switching to using our own assertion library (`go.mongodb.org/mongo-driver/internal/testutil/assert`). The syntax is very similar but the assertions require a message to go along with them:

```
assert.Nil(t, err, "UnmarshalExtJSON error: %v", err)
assert.Equal(t, "value", responseDoc.DefinedField, "expected DefinedField to be 'value', got %q", responseDoc.DefinedField)
```

This makes them more verbose, but they're clearer because, unlike `testfiy`, they don't rely on the "expected" and "actual" values to be passed in a specific order.

**Review by divjotarora — COMMENTED** [review]
This looks fantastic! The code changes all LGTM and I just have a couple of minor comments on the tests. I'll hold off on approving until those are done, but great work!

File: bson/unmarshal_test.go (hunk 8 above)
Comment (benjirewis) [reply bson/unmarshal_test.go:175]: That sounds good! Moved them within `TestUnmarshalExtJSONWithUndefinedField()`, added the references to `testing.T` and changed the name.

File: bson/unmarshal_test.go (hunk 9 above)
Comment (benjirewis) [reply bson/unmarshal_test.go:276]: Ah gotcha, thanks for the explanation. Switched to use `testutil/assert`.

**Review by divjotarora — APPROVED** [review]
LGTM. Great work!
